src/devices/StepperMotor/StepperMotor.cs

```diff
@@ -0,0 +1,130 @@
+        /// <param name="pin2">The GPIO pin number which corresponds pin B on ULN2003 driver board.</param>
+        /// <param name="pin3">The GPIO pin number which corresponds pin C on ULN2003 driver board.</param>
+        /// <param name="pin4">The GPIO pin number which corresponds pin D on ULN2003 driver board.</param>
+        /// <param name="mode">The stepper's mode.</param>
+        /// <param name="rpm">Revolutions per minute.</param>
+        public StepperMotor(int pin1, int pin2, int pin3, int pin4, StepperMode mode = StepperMode.HalfStep, uint rpm = 15)
+        {
+            _pin1 = pin1;
+            _pin2 = pin2;
+            _pin3 = pin3;
+            _pin4 = pin4;
+            _mode = mode;
+            _rpm = rpm > 0 ? rpm : 15; // Default revolutions per minute for 28BYJ-48 is approximately 15
+
+            _controller = new GpioController();
+
+            _controller.OpenPin(_pin1, PinMode.Output);
+            _controller.OpenPin(_pin2, PinMode.Output);
+            _controller.OpenPin(_pin3, PinMode.Output);
+            _controller.OpenPin(_pin4, PinMode.Output);
+        }
+
+        /// <summary>
+        /// Stop the motor.
+        /// </summary>
+        public void Stop()
+        {
+            _steps = 0;
+            _stopwatch.Stop();
+            _controller.Write(_pin1, PinValue.Low);
+            _controller.Write(_pin2, PinValue.Low);
+            _controller.Write(_pin3, PinValue.Low);
+            _controller.Write(_pin4, PinValue.Low);
+        }
+
+        /// <summary>
+        /// Moves the motor. If the number is negative, the motor moves in the reverse direction.
+        /// </summary>
+        /// <param name="steps">Number of steps.</param>
+        public void Step(int steps)
+        {
+            double lastStepTime = 0;
+            _stopwatch = Stopwatch.StartNew();
+            _isClockwise = steps >= 0;
+            _steps = Math.Abs(steps);
+            stepMicrosecondsDelay = 60 * 1000 * 1000 / _steps / _rpm;
+            _currentStep = 0;
+
+            while (_currentStep < _steps)
+            {
+                double elapsedMicroseconds = _stopwatch.Elapsed.TotalMilliseconds * 1000;
+
+                if (elapsedMicroseconds - lastStepTime >= stepMicrosecondsDelay)
+                {
+                    lastStepTime = elapsedMicroseconds;
+
+                    if (_isClockwise)
+                        _engineStep = _engineStep - 1 < 1 ? 8 : _engineStep - 1;
+                    else
+                        _engineStep = _engineStep + 1 > 8 ? 1 : _engineStep + 1;
+
+                    ApplyEngineStep();
+                    _currentStep++;
+                }
+            }
+        }
+
+        private void ApplyEngineStep()
+        {
+            if (_mode == StepperMode.HalfStep)
+            {
+                _controller.Write(_pin1, _engineStep == 8 || _engineStep == 1 || _engineStep == 2);
```

Comment:
you can define a static array for those, i.e.:
`static bool[] s_halfStep = new bool[8] { true, true, true, false, false, false, false, false }`
- change step to be numbered from 0 and then:

```csharp
_controller.Write(_pin1, s_halfStep[_step]);
_controller.Write(_pin2, s_halfStep[(_step + 2) % 8]);
_controller.Write(_pin3, s_halfStep[(_step + 4) % 8]);
_controller.Write(_pin3, s_halfStep[(_step + 6) % 8]);
```

etc.. that should be doable for all modes I believe.

The steps increment can be done like that:

```csharp
// somewhere have
int inc = 1; // or -1 depending on direction
_step = (_step + inc + 8) % 8;
```

Reply:
Got it! To be honest the code above is not worked as expected for me :-)

Krzysztof, I use the following sequence for halfstep mode :
![Sequence](http://www.geeetech.com/wiki/images/6/6b/Stepper_sequence.jpg)

So for this mode I've added appropriate array:

        static bool[,] _halfStepSequence = new bool[4, 8] {
                 { true, true, false, false, false, false, false, true },
                 { false, true, true, true, false, false, false, false },
                 { false, false, false, true, true, true, false, false },
                 { false, false, false, false, false, true, true, true }
            };
I think this is more obvious. What do you think?

Reply:
That is fine too but I think the code will work once you change _step to be 0..7 instead of 1..8 (note that swapping step 8 with 0 and putting it as a first column is equivalent will make the table look more obvious)

Reply:
I think the formula would only work for the half step and we would need to think how to write it in generic way to work for other types

Reply:
Thanks! Generic way is good idea. Now I've added 3 matrix for each stepper motor types, not only for half step. Please find these changes in the 3rd commit